Implementation of $S^1$ model

robertjankowski opened the pull request:
This is an implementation of $\mathbb{S}^1$ model by [1]. It is a geometric soft configuration model which produces synthetic networks with similar properties as in real networks, i.e., high level of clustering, heterogeneous degree distribution, or small-world property.


[1] Serrano, M. Angeles, Dmitri, Krioukov, and Marián, Boguñá. "Self-Similarity of Complex Networks and Hidden Metric Spaces". Phys. Rev. Lett. 100 (2008): 078701.

robertjankowski commented:
Hi! Can I ask for the code review? @MridulS

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+    The $\mathbb{S}^1$ model is the simplest among the class of geometric models [1].
+    The similarity space is a one dimensional sphere—a circle of radius R—where N nodes are
+    distributed with a fixed density, set to one without loss of generality, so that $N = 2\pi R$.
+    Each node is also given a hidden variable ``'kappa'`` proportional to its expected degree.
```
```suggestion
    Each node is also given a hidden variable ``kappa`` proportional to its expected degree.
```

Just some cosmetic changes here, we already have `kappa` in a code block no need to add it inside quotes :)
Could you also do remove these quotes from all the other codeblocks too? Thanks!

You can check how this doc is rendered here https://output.circle-artifacts.com/output/job/6af83fc3-9162-4cab-9fab-75baa39c2fd9/artifacts/0/doc/build/html/reference/generated/networkx.generators.geometric.S1_graph.html#s1-graph

robertjankowski replied:
Sure! I removed the quotes

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+    .. [1] Serrano, M. Angeles, Dmitri, Krioukov, and Marián, Boguñá. "Self-Similarity
+    of Complex Networks and Hidden Metric Spaces". Phys. Rev. Lett. 100 (2008): 078701.
+    """
+    assert beta > 1
```
We should raise a more informative error here. If the user passes in `beta` less than 1 it will just give an assertion error. We should also raise an error like `The inverse temperature can not be less than 1` or something like that :)

robertjankowski replied:
I extended the implementation for the case when beta<1. And I am now raising the exception when beta<0.

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+
+
+@py_random_state(5)
+def S1_graph(n, beta, gamma, mean_degree, kappas=None, seed=None):
```
```suggestion
def S1_graph(n, * ,  beta=None, gamma=None, mean_degree=None, kappas=None, seed=None):
```

We shouldn't force users to provide `beta`, `gamma`, `mean_degree` when they are already providing `kappas`. This signature will also force users to be explicit about the parameters they are passing. So they would need to call this function with something like `nx.S1_graph(100, beta=1.5, gamma=2.7, mean_degree=5)` or `nx.S1_graph(100, kappas=kappas_dict)`.

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+        Exponent of powerlaw distribution of the hidden degrees
+    mean_degree: float
+        The mean degree of the network
+    kappas: list, optional
+        Values of hidden degrees for each node
```
Maybe it makes sense for `kappas` to be a dictionary with node as the key and kappa as the value?

robertjankowski replied:
For now, I wanted to keep the implementation simple, but if you think that using dict will be better, I can extend it.

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+    """
+    assert beta > 1
+
+    if kappas is not None:
```
```suggestion
    if kappas is not None:
        if not all((beta is None, gamma is None, mean_degree is None)):
             raise NetworkXError('..........')
```

Maybe something like this so it explicitly catches the cases where a user has provided `kappas` and also the other variables?

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+def S1_graph(n, beta, gamma, mean_degree, kappas=None, seed=None):
+    r"""Returns a $\mathbb{S}^1$ model.
+
+    The $\mathbb{S}^1$ model is the simplest among the class of geometric models [1].
```
```suggestion
    The $\mathbb{S}^1$ model is the simplest among the class of geometric models [1]_.
```

It needs to have a `_` in the end to properly cite the reference below :)

MridulS commented on networkx/generators/geometric.py:
```diff
@@ -784,3 +785,110 @@ def thresholded_random_geometric_graph(
+
+    Examples
+    --------
+    Default Graph:
+
+    G = nx.S1_graph(100, 1.5, 2.7, 5)
+
+    Custom Graph:
+
+    Create a $\mathbb{S}^1$ model with 100 nodes. The $\beta$ parameter is set to 1.5
+    and the exponent of the powerlaw distribution of the hidden degrees is 2.7
+    with mean value of 5.
```
The example should have something that is quickly copy-able so we usually put in code blocks here. Could you move the examples from the `Notes` section with the python`>>>` identifiers to examples?

MridulS commented on networkx/generators/tests/test_geometric.py:
```diff
@@ -329,3 +329,47 @@ def test_geometric_edges_raises_no_pos():
         nx.geometric_edges(G, radius=1)
+
+
+class TestS1Graph:
```
No need to create a class structure here :)

I know we have classes all across our tests but we are trying to move away from them.

robertjankowski replied:
I removed the class there

MridulS reviewed (commented):
Thank you for this PR @robertjankowski! It looks good!

I have made some comments about code/API organization in the review. Let me know what do you think.

Also I think we can speed up a bunch of calculations with numpy vectorisation but it's not a major blocker here. It's fine to keep this graph generator with vanilla python :)

robertjankowski commented:
Thank you very much for the review @MridulS. Let me know if some more changes are needed :)

robertjankowski commented:
Hi @MridulS, is the PR good to merge now?

robertjankowski commented:
Hi @MridulS, do you have any further comments?

dschult reviewed (commented):
Thanks for this. 
It'd be good to have more language introducing the user to what this means and what the literature says about it, or why someone might use it. You use the term "Geometric Model", but that can mean so many different things, and it isn't easy to look it up and find the network stuff amidst all the other types of geometric models. You also talk about a similarity space but don't say anything about what that means. S1 is often used to denote the unit circle, but how that relates to a network is not clear either. It would make me happy to see a paragraph or two in a notes section introducing the user to this concept and shy it might be useful. Then make sure the notation and words you are using in the start of the doc_string are all defined somewhere in the doc_string (or give a general reference like wikipedia, for people to read to find out more.

Thoughts?

I put some more comments below -- hopefully useful.

robertjankowski commented:
Thank you @dschult for your comments! 

I added more explanations about the model in docs, changed `kappa` to dictionary type, and updated the test cases. Regarding the function's name, in the literature, the S1 model is widely used so I would be inclined to keep this name. 

Let me know if some more changes are needed :)

dschult reviewed (commented):
Thanks for these changes! 

I'm still concerned that the name `S1_model` doesn't fit well with the networkx culture of function names.  (see more below in the results section of the doc_string)

Is there a relationship between this and the `soft_random_geometric_graph`? It seems like this is that with dim=1, distance measured on a circle instead of euclidean, and a specific p_dist function that depends on randomly assigned node traits. Actually, the fact that the probabilities depend on node traits means this isn't a strict geometric model is it... :).  

Minor nit-picky things: 
- could you wrap the doc_string similar to the code width length of 80 chars or so (code should be <88 chars, so something close to 80 works well).
- the tests should use a seed value so we don't get a failure that we can't reproduce. To be clear, ideally the test would always pass -- and it wouldn't matter which seed value we choose. But we can't test all possible seeds. So we should probably (not in the test suite) run the test for lots of different seed values. But in the test suite we should set a seed, test that seed and use that as a measure of whether the code is working as we expect in that case. We want to avoid the maintenance headache for a rare failure that can't be reproduced.

robertjankowski commented:
Thanks again for the review!

I implemented all your suggestions. Below are some comments to your inquiries:

The S^1 model can be also described as a purely geometric H^2 model where nodes lie in the hyperbolic plane. We need to map the hidden degrees into the radial coordinates. By doing that, the hubs will be located closer to the center of the hyperbolic planes, whereas degree-one nodes are close to the border.

I changed the name of the function. I like your suggestion (soft_random_S1_geometric_graph), but since we can also describe it using H^2, I decided to name it as soft_random_S1_H2_geometric_graph.

—

Regarding the relation to the gravity low, it is a more intuitive interpretation, i.e., nodes with higher hidden degrees (the equivalent of mass) will be more likely to be connected. Additionally, the similarity (the angular distance) is related to the r of the gravity low. In this case, nodes closer in the similarity space have more probability to form a connection. 

Once we map the connection probability of the S^1 model into H^2, we can see that it is nothing but the Fermi-Dirac distribution. It appears because we allow only one link between a pair of nodes. So, the hyperbolic distances can now be interpreted as energies of fermionic links, whereas hyperbolic disk radius R is the chemical potential, and beta = 1/T is the inverse temperature.

Yes, I was referring to the clustering coefficient. I updated the docs.

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,206 @@ def thresholded_random_geometric_graph(
+    adjusted to $R = N/2\pi$,  where $N$ is the number of nodes, so that the density
+    is set to 1 without loss of generality.
+
+    The connection probability between any pair of nodes takes the form
+    of a gravity law, whose magnitude increases with the product of the
+    hidden degrees (i.e., their combined popularities), and decreases with
+    the angular distance between the two nodes.
```
Maybe something like:
```
    The connection probability pairs of nodes is similar to a gravity law in that
    hidden degree acts like mass (causing attraction) and that attraction decreases
    with angular distance (though only linearly unlike gravity's square dependence).
```
Maybe the gravity bit is a distraction -- explaining the increase and decrease might be enough, but you can decide.

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,206 @@ def thresholded_random_geometric_graph(
+    n, gamma, mean_degree (if provided) are used to construct random kappas
+    sampled from the power-law distribution.
+
+    beta: float
```
```suggestion
    beta: positive number
```

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,206 @@ def thresholded_random_geometric_graph(
+
+    beta: float
+        Inverse temperature, controlling the clustering coefficient.
+    n: int, optional
```
For the optional arguments, add `(default: <default value>)` to the type description in the docs.  No word "optional" is needed if there is a default.

dschult reviewed (commented):
Ahh... thanks for the connection to H2 as well as S1. It gives be a better idea of what is crucial to the arrangement. It isn't specific to S1, or to H2.  There are many other spaces that this geometry could be embedded inside. The crux is that the distance measure is periodic.  How about the name: `soft_random_periodic_geometric_graph`. 
That is getting a little long, so maybe `soft_periodic_geometric_graph`?  But sometimes brevity causes confusion.

robertjankowski commented:
Thank you for your comments!

I checked with the literature, and this kind of model is called the geometric soft configuration model. Hence, I propose to use it as a function name, i.e., `geometric_soft_configuration_graph`. What do you think?

I made `beta` as a keyword parameter and changed a few things in the doc string.

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,205 @@ def thresholded_random_geometric_graph(
+    radii = {}
+    zeta = 1 if beta > 1 else 1 / beta
+    kappa_0 = min(list(kappas.values()))
+    R_hat = (2 / zeta) * math.log(n / math.pi) - (2 * max(1, beta)) / (
+        beta * zeta
+    ) * math.log(mu * kappa_0**2)
+
+    for k, kappa in kappas.items():
+        radii[k] = R_hat - (2 * max(1, beta)) / (beta * zeta) * math.log(
+            kappa / kappa_0
+        )
```
Here's an attempt to improve readability (I used R_c for curvature, but you may have a better name):
```python
    R_c = 2 * max(1, beta) / (beta * zeta)
    R_hat = (2 / zeta) * math.log(n / math.pi) - R_c * math.log(mu * kappa_0**2)
    R_max = R_hat + R_c * math.log(kappa_0)

    for node, kappa in kappas.items():
        radii[node] = R_max - R_c * math.log(kappa)
```

You could also replace the loop with a dict-comprehension like:
```python
    radii = {node: R_max - R_c * math.log(kappa) for node, kappa in kappas.items()}
```

robertjankowski replied:
Yeah, I would go with removing the square term and simplifying a bit.

dschult commented on networkx/generators/tests/test_geometric.py:
```diff
@@ -333,3 +333,89 @@ def test_geometric_edges_raises_no_pos():
+    G = nx.geometric_soft_configuration_graph(
+        beta=1.5, n=100, gamma=2.7, mean_degree=10, seed=42
+    )
+    kappas = list(nx.get_node_attributes(G, "kappa").values())
+    assert len(kappas) == 100
+    thetas = list(nx.get_node_attributes(G, "theta").values())
+    assert len(thetas) == 100
```
```suggestion
    kappas = nx.get_node_attributes(G, "kappa")
    assert len(kappas) == 100
    thetas = nx.get_node_attributes(G, "theta")
    assert len(thetas) == 100
    radii = nx.get_node_attributes(G, "radius")
    assert len(radii) == 100
```

dschult commented on networkx/generators/tests/test_geometric.py:
```diff
@@ -333,3 +333,89 @@ def test_geometric_edges_raises_no_pos():
+    G = nx.geometric_soft_configuration_graph(
+        beta=2.5, n=5000, gamma=2.7, mean_degree=10, seed=42
+    )
+    kappas = list(nx.get_node_attributes(G, "kappa").values())
+    mean_kappas = sum(kappas) / len(kappas)
```
```suggestion
    kappas = nx.get_node_attributes(G, "kappa")
    mean_kappas = sum(kappas.values()) / len(kappas)
```

dschult commented on networkx/generators/tests/test_geometric.py:
```diff
@@ -333,3 +333,89 @@ def test_geometric_edges_raises_no_pos():
+    kappas = {i: 10 for i in range(1000)}
+    G = nx.geometric_soft_configuration_graph(beta=1, kappas=kappas)
+    assert len(G) == 1000
+    kappas = list(nx.get_node_attributes(G, "kappa").values())
+    mean_kappas = sum(kappas) / len(kappas)
+    assert mean_kappas == 10
```
```suggestion
    kappas = nx.get_node_attributes(G, "kappa")
    assert all(kappa == 10 for kappa in kappas.values())
```

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,205 @@ def thresholded_random_geometric_graph(
+def geometric_soft_configuration_graph(
+    *, beta, n=None, gamma=None, mean_degree=None, kappas=None, seed=None
+):
+    r"""Returns a $\mathbb{S}^1/\mathbb{H}^2$ model.
```
```suggestion
    r"""Returns a random graph from the geometric soft configuration model
```

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,205 @@ def thresholded_random_geometric_graph(
+
+    # Generate random positions on a circle
+    thetas = {k: seed.uniform(0, 2 * math.pi) for k in kappas}
+    mapping = dict(zip(range(n), kappas.keys()))
+
+    for i in range(n):
+        for j in range(i):
+            k, l = mapping[i], mapping[j]
```
It's traditional (at least in networkx) to denote pairs of nodes using `u` and `v` rather than `i, j, k, l`. Also, we don't need to convert back and forth between integers and nodes. What do you think of:
```
    G = nx.Graph()
    for u in kappas:
        for v in list(G):
            <compute p_ij using u and v instead of k and l>
            if seed.random() < p_ij:
                G.add_edge(u, v)
        G.add_node(u)  # ensures u becomes a node in G
```
We need the `list(G)` because we may add the node u during this loop.
And since no edges might be added, we put the `add_node` after the loop over `v`.

**silliness:**  Another possible convenience is that $\tau$ is defined to be $2*\pi$. And python has `math.tau` in addition to `math.pi`. I guess there is a long-standing debate about whether $\tau$ or $\pi$ should be the dominant circle constant taught in school.  So, if you want to expand the reputation of `tau` you can use `math.tau` wherever you have `2 * math.pi`.

dschult reviewed (commented):
Nice Name!  Thanks for coming up with that. :}

This review is getting into the weeds a little, but I think that means we are getting near the end.  :)

It strikes me that kappa is always created by multiplying by kappa_0 and then used by dividing by kappa_0.  Could we just scale out the kappa_0?  Also the kappa_0 used when creating it is from a formula, while when dividing by it it seems to be the minimum value of all kappas. So, I might be getting confused by all that. But it seems that we could scale all the kappas by kappa_0 at some poin and reduce a lot of computation.

I was surprised to find that `(kappa_c / kappa_0) ** (1 - gamma)` is just `1 / n`.  Is this really right? or did I do some algebra mistake... That simplifies formulas quite a bit. I'm used to published formulas already being simplified. Are these from the publication?

It'd be great if we had test(s) which tested values. We've got a few checking e.g. that increasing beta decreases average clustering, and that beta<1 gives positive average clustering.  Are there others?  Perhaps with average degree while changing either the kappas or mean_degree?

robertjankowski commented:
Thanks @dschult for the review! I think now it looks better.

Regarding the kappas you're right and we can simplify it. The equation I used allowed us to change the kappa cutoff (kappa_c), but in our case, it is defined using kappa_0. 

I added a few more tests to check the connectivness and mean degree of the generated networks.

I hope we are close to finishing this PR :)

robertjankowski commented:
Hi @dschult, do you have any further comments?

robertjankowski commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,205 @@ def thresholded_random_geometric_graph(
+    variables: a hidden degree $\kappa_i$, quantifying its popularity, influance,
+    or importance, and an angular position $\theta_i$ in a one-dimensional
+    sphere (or circle) abstracting the similarity space, where angular distances
+    between nodes are a proxy for their similarity. The radius of the circle is
```
It's fine, I just replaced sometimes with often since this name is very popular in the literature.

robertjankowski commented:
Thanks @dschult! I think this will be the last round :)

dschult reviewed (approved):
This looks good to me!   Thanks @robertjankowski !!

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,200 @@ def thresholded_random_geometric_graph(
     return G
+
+
+@py_random_state(5)
```
New decorator for dispatching to bacend packages:
```suggestion
@py_random_state(5)
@nx._dispatch(graphs=None)
```

dschult commented on networkx/generators/geometric.py:
```diff
@@ -844,3 +845,200 @@ def thresholded_random_geometric_graph(
+    clustering, and self-similarity.
+
+    In the geometric soft configuration model, a node $i$ is assigned two hidden
+    variables: a hidden degree $\kappa_i$, quantifying its popularity, influance,
```
Spelling:
```suggestion
    variables: a hidden degree $\kappa_i$, quantifying its popularity, influence,
```

dschult reviewed (commented):
minor tweaks -- spelling and add dispatch decorator support.

rossbar commented:
Sounds reasonable for me, thanks for taking the time to put together the summary!
